### PR TITLE
[Translation][File dumper] allow get file content without writing in file.

### PR DESCRIPTION
**Pull request by aitboudad**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets | #14881 |
| Tests pass? | yes |
| License | MIT |

### CODE REVIEW

File: src/Symfony/Component/Translation/Dumper/FileDumper.php [hunk 1]
```diff
@@ -63,26 +63,34 @@ public function setBackup($backup)
      */
     public function dump(MessageCatalogue $messages, $options = array())
     {
-        if (!array_key_exists('path', $options)) {
-            throw new \InvalidArgumentException('The file dumper needs a path option.');
+        if (!array_key_exists('path', $options) && !array_key_exists('format_callback', $options)) {
```
Comment (stof) [inline src/Symfony/Component/Translation/Dumper/FileDumper.php]: the name format_callback looks weird to me. This callback is not responsible for any formatting, but for saving

File: src/Symfony/Component/Translation/Tests/Dumper/FileDumperTest.php [hunk 2]
```diff
@@ -16,6 +16,20 @@
...
 {
+    public function testDumpWithFormatCallback()
+    {
+        $catalogue = new MessageCatalogue('en');
+        $catalogue->add(array('foo' => 'bar'));
+
+        $dumper = new ConcreteFileDumper();
+        $format = false;
+        $dumper->dump($catalogue, array('format_callback' => function ($content) use (&$format) {
+            $format = true;
```
Comment (stof) [inline src/Symfony/Component/Translation/Tests/Dumper/FileDumperTest.php]: you should probably have an assertion on the content too, not just on the fact it got called

File: src/Symfony/Component/Translation/Dumper/FileDumper.php (hunk 1 above)
Comment (aitboudad) [inline src/Symfony/Component/Translation/Dumper/FileDumper.php]: what's about `save_handler` ?

File: src/Symfony/Component/Translation/Tests/Dumper/FileDumperTest.php (hunk 2 above)
Comment (aitboudad) [inline src/Symfony/Component/Translation/Tests/Dumper/FileDumperTest.php]: fixed

File: src/Symfony/Component/Translation/CHANGELOG.md [hunk 3]
```diff
@@ -5,8 +5,9 @@ CHANGELOG
 -----
 
  * deprecated Translator::getMessages(), rely on TranslatorBagInterface::getCatalogue() instead.
- * added option `json_encoding` to JsonFileDumper
- * added options `as_tree`, `inline` to YamlFileDumper
+ * added option `save_handler` to FileDumper.
+ * added option `json_encoding` to JsonFileDumper.
+ * added options `as_tree`, `inline` to YamlFileDumper.
```
Comment (fabpot) [inline src/Symfony/Component/Translation/CHANGELOG.md]: The three dots should be removed

Comment (aitboudad) [inline src/Symfony/Component/Translation/CHANGELOG.md]: fixed

**Comment by aitboudad** [comment]
ping @symfony/deciders

**Comment by Tobion** [comment]
This looks like a workaround for a bad architecture. It doesn't make sense that the formating is tied to dumping to files at all. We could alternatively refactor out the formatting part of the FileDumpers.

So basically refactor the `formatCatalogue` out of the dumpers into a Formatter for each type. Then the file dumper would just accept a formatter and put in content in a file. And people who do not want to save the content to a file would just simply use the formatter alone.

**Comment by aitboudad** [comment]
@Tobion thanks for pointing me in the right direction. Seems making `formatCatalogue` public is enough, can you review it now ?

File: src/Symfony/Component/Translation/Tests/Dumper/IniFileDumperTest.php [hunk 4]
```diff
@@ -16,17 +16,13 @@
...
         $catalogue->add(array('foo' => 'bar'));
 
-        $tempDir = sys_get_temp_dir();
         $dumper = new IniFileDumper();
-        $dumper->dump($catalogue, array('path' => $tempDir));
 
-        $this->assertEquals(file_get_contents(__DIR__.'/../fixtures/resources.ini'), file_get_contents($tempDir.'/messages.en.ini'));
-
-        unlink($tempDir.'/messages.en.ini');
+        $this->assertEquals(file_get_contents(__DIR__.'/../fixtures/resources.ini'), $dumper->formatCatalogue($catalogue, 'messages'));
```
Comment (Tobion) [inline src/Symfony/Component/Translation/Tests/Dumper/IniFileDumperTest.php]: `assertStringEqualsFile`?!

**Comment by Tobion** [comment]
I see `format` is deprecated since 2.8. IMO the `abstract` should be removed from it, so people do not have to implement the deprecated method.

File: src/Symfony/Component/Translation/Tests/Dumper/IniFileDumperTest.php (hunk 4 above)
Comment (aitboudad) [inline src/Symfony/Component/Translation/Tests/Dumper/IniFileDumperTest.php]: done,  didn't know it :)

**Comment by aitboudad** [comment]
yes see #15832

**Comment by Tobion** [comment]
It don't like that formatting and dumping is still in one class. But this way it might be good enough :+1:

**Comment by fabpot** [comment]
Thank you @aitboudad.

**Pull request merged**